core/src/main/java/com/linecorp/armeria/client/endpoint/RingHashEndpointSelectionStrategy.java

```diff
@@ -0,0 +1,205 @@
+                weightedRingEndpoint = new WeightedRingEndpoint(endpoints), true
+            );
+        }
+
+        @Override
+        public Endpoint selectNow(ClientRequestContext ctx) {
+            final WeightedRingEndpoint weightedRingEndpoint = this.weightedRingEndpoint;
+            if (weightedRingEndpoint == null) {
+                // 'endpointGroup' has not been initialized yet.
+                return null;
+            }
+            return weightedRingEndpoint.select(ctx.endpoint());
+        }
+        
+        private final class WeightedRingEndpoint {
+            private final Int2ObjectSortedMap<Endpoint> ring = new Int2ObjectAVLTreeMap<>();
+            private final Iterable<Endpoint> endpoints;
+
+            Endpoint select(Endpoint point) {
+                final int key = getXXHash(point.host());
+                final SortedMap<Integer, Endpoint> tailMap = ring.tailMap(key);
+                return tailMap.isEmpty() ? ring.get(ring.firstKey()) : tailMap.get(tailMap.firstKey());
+            }
+
+            WeightedRingEndpoint(Iterable<Endpoint> endpoints) {
+                // prepare immutable endpoints
+                this.endpoints = Streams.stream(endpoints)
+                                        .filter(e -> e.weight() > 0) // only process endpoint with weight > 0
+                                        .sorted(Comparator.comparing(Endpoint::weight)
+                                                          .thenComparing(Endpoint::host)
+                                                          .thenComparingInt(Endpoint::port))
+                                        .collect(toImmutableList());
+
+                // The GCD is properly sized so that it does not exceed the size of the ring
+                final int gcd = findGcdInEndpoints(this.endpoints);
+                final int numberOfEndpointInTheRing = caculateNumberOfEndpointInTheRing(this.endpoints, gcd);
+                final int sizeOfRing = getSize(this.endpoints);
+                if (sizeOfRing >= numberOfEndpointInTheRing) {
+                    for (Endpoint endpoint : this.endpoints) {
+                        final int weight = endpoint.weight();
+                        final String host = endpoint.host();
+                        // If weight is 3 and gcd is 1, place 3 times in the ring
+                        // if weight is 1 and gcd is 1, place once in the ring
+                        // If weight is 3 and gcd is 3, place 1 times in the ring
+                        final int count = weight / gcd;
+                        for (int i = 0; i < count; i++) {
+                            final String weightedHost = host + weight;
+                            final int hash = getXXHash(weightedHost);
+                            ring.put(hash, endpoint);
+                        }
+                    }
+                }
+
+                // When the size of the GCD is too small and exceeds the size of the ring,
```

Comment:
I don't see `the size of the ring` is a reasonable limitation for the maximum size of the ring. Some users might want to use more memory to get exact weighted balancing. 

What do you think of taking the maximum size of the ring as an input parameter?